### PR TITLE
Check polynomial degree of exactness of quadrature

### PR DESCRIPTION
**Pull request by mbaudin47**
Based on a suggestion by J.Peter (ONERA), here are unit tests of quadrature based on their polynomial degree of exactness.
- `GaussProductExperiment`
- `SmolyakExperiment`
- `TensorProductExperiment`
- `GaussLegendre`

Update documentation of polynomial degree of exactness of the same classes.

### CODE REVIEW

**Comment by adutfoy** [comment]
Michaël,
Here are some minor remarks.
PRQuad-1

- $a_i < b_i$ and not $a_i \leq b_i$ because other wose the integral = 0
- why not call the number od node $n_i$ rather than $s_i$?
- numbers and not number
- $w_{k_i}^{(i)}>0$ and not $w_{k_i}^{(i)}\geq 0$

**Comment by adutfoy** [comment]
You should introduce better that this is another example. For example you could write: 
We show how to use this method to evaluate the following integral:
PRQuad-2

First, we use the default number of nodes as defined in the Resourcemap class:   you give the pyhton lines you have already written
..
Now, we set the number of nodes....

**Comment by adutfoy** [comment]
I continue my review... About SmolyakExperiment: 

- you should explain that the sets of nodes are nested sets so that we understand better the $\Delta_k^{(1)}$ set.
- section 'Merge duplicate node': you announced $(x_1, x_2)$ points to be merged and thnen you used $(x_1, y_1)$
- I would have written
$T_\ell^{(d)} = Q_{\ell,1}^{(1)} \times \dots \times Q_{\ell,d}^{(1)} $ 
to distinguish the quadratures according to the components

**Comment by adutfoy** [comment]
A last question about the integral we evaluate with the Smolyak quadrature: must the set on which we integrate $\chi$ be a multivariate interval?

File: python/src/SmolyakExperiment_doc.i.in [hunk 1]
```diff
@@ -195,21 +194,49 @@ If the bounds of the input distribution are either very close to zero
...
+
+    \bigotimes_{i = 1}^\inputDim \mathcal{P}_{m_i}^{(1)}
+    = 
+    \left\{
+    (x_1, ..., x_\inputDim)\in\Rset^\inputDim
+    \rightarrow \prod_{i = 1}^\inputDim p_i(x_i) \in \Rset, \quad 
+    p_i \in \mathcal{P}_{m_i}^{(1)}
+    \right\}.
+
+Therefore the Smolyak quadrature is exact for all polynomials of the non-classical
```
Comment (mbaudin47) [inline python/src/SmolyakExperiment_doc.i.in:214]:
```suggestion
Assume that :math:`m_{k_i}` is the polynomial degree of exactness of the :math:`i`-th marginal quadrature rule, for any math:`i \in \{1,...,\inputDim\}`.
Therefore the Smolyak quadrature is exact for all polynomials of the non-classical 
```

**Comment by mbaudin47** [comment]
@regislebrun , @adutfoy , @josephmure : Can you please review this PR?

File: python/src/GaussLegendre_doc.i.in [hunk 2]
```diff
@@ -9,34 +9,133 @@ Available constructors:
...
-with :math:`f: \Rset^d \mapsto \Rset^p`, :math:`\vect{a}, \vect{b}\in\Rset^d` using a fixed tensorized Gauss-Legendre approximation:
+where :math:`\model: \Rset^\inputDim \mapsto \Rset^\outputDim` is a function and
+:math:`\vect{a}, \vect{b}\in\Rset^\inputDim` are the bounds
+where :math:`a_i \leq b_i` for :math:`i \in \{1, ..., \inputDim\}`.
+For :math:`i \in \{1, ..., \inputDim\}`, let :math:`s_i \in \Nset` be the number
+of marginal nodes of the univariate Gauss-Legendre quadrature.
+The algorithm uses a fixed tensorized Gauss-Legendre approximation based
+on the tensor-product Gauss quadrature using
+:math:`s_1, ..., s_\inputDim \in \Nset` marginal nodes.
+In other words, this algorithms uses an anisotropic quadrature rule,
```
Comment (jschueller) [inline python/src/GaussLegendre_doc.i.in:36]: singular

File: python/src/SmolyakExperiment_doc.i.in [hunk 3]
```diff
@@ -195,21 +194,50 @@ If the bounds of the input distribution are either very close to zero
...
+
+    \bigotimes_{i = 1}^\inputDim \mathcal{P}_{m_i}^{(1)}
+    = 
+    \left\{
+    (x_1, ..., x_\inputDim)\in\Rset^\inputDim
+    \rightarrow \prod_{i = 1}^\inputDim p_i(x_i) \in \Rset, \quad 
+    p_i \in \mathcal{P}_{m_i}^{(1)}
+    \right\}.
+
+Assume that :math:`m_{k_i}` is the polynomial degree of exactness of the :math:`i`-th marginal univariate quadrature rule using :math:`k_i` nodes, for any math:`i \in \{1,...,\inputDim\}`.
```
Comment (jschueller) [inline python/src/SmolyakExperiment_doc.i.in:214]: line too long

File: python/test/t_GaussLegendre_std.py [hunk 4]
```diff
@@ -19,3 +20,95 @@
...
+    absoluteError = abs(computedIntegral - expectedIntegral)
+    if verbose:
+        print(
+            f"Polynomial : {str(productPoly):20s}, "
+            f"  integral computed  = {computedIntegral:.7f}, "
+            f"  expected = {expectedIntegral:.7f}, "
+            f"  absolute error = {absoluteError:.3e}"
+        )
+    ott.assert_almost_equal(expectedIntegral, computedIntegral, rtol, atol)
+    return
```
Comment (jschueller) [inline python/test/t_GaussLegendre_std.py:104]: useless return

File: python/test/t_SmolyakExperiment_std.py [hunk 5]
```diff
@@ -306,3 +306,105 @@ def testSmolyakExperiment5():
...
+    absoluteError = abs(computedIntegral - expectedIntegral)
+    if verbose:
+        print(
+            f"Polynomial : {str(productPoly):20s}, "
+            f"  integral computed  = {computedIntegral:.7f}, "
+            f"  expected = {expectedIntegral:.7f}, "
+            f"  absolute error = {absoluteError:.3e}"
+        )
+    ott.assert_almost_equal(expectedIntegral, computedIntegral, rtol, atol)
+    return
```
Comment (jschueller) [inline python/test/t_SmolyakExperiment_std.py:397]: there too

**Review by jschueller — APPROVED** [review]

**Review by regislebrun — APPROVED** [review]

**Comment by mbaudin47** [comment]
> Michaël, Here are some minor remarks. 
>     * $a_i < b_i$ and not $a_i \leq b_i$ because otherwise the integral = 0

Actually, the code does not prevent from using a zero-volume interval: https://github.com/openturns/openturns/blob/39d39f0dc9b1a6e37edeca15bef568643400a33c/lib/src/Base/Algo/GaussLegendre.cxx#L84. In this case, it just returns zero.

File: python/test/t_SmolyakExperiment_std.py (hunk 5 above)
Comment (mbaudin47) [reply python/test/t_SmolyakExperiment_std.py]: Ok. Fixed at 2e5c324cbafb145a0021ee382d5c91fd1fe3f579

File: python/test/t_GaussLegendre_std.py (hunk 4 above)
Comment (mbaudin47) [reply python/test/t_GaussLegendre_std.py]: Ok. Fixed at 2e5c324cbafb145a0021ee382d5c91fd1fe3f579

File: python/src/SmolyakExperiment_doc.i.in (hunk 3 above)
Comment (mbaudin47) [reply python/src/SmolyakExperiment_doc.i.in]: Ok. Fixed at 2e5c324cbafb145a0021ee382d5c91fd1fe3f579

File: python/src/GaussLegendre_doc.i.in (hunk 2 above)
Comment (mbaudin47) [reply python/src/GaussLegendre_doc.i.in]: Ok. Fixed at 2e5c324cbafb145a0021ee382d5c91fd1fe3f579

**Comment by mbaudin47** [comment]
> I continue my review... About SmolyakExperiment:
> 
>     * you should explain that the sets of nodes are nested sets so that we understand better the  Δk(1) set.

This is not mandatory. Even if the marginal quadratures are not nested, the Smolyak quadrature produce a valid quadrature rule.

>     * I could have written Tℓ(d)=... to distinguish the quadratures according to the components

This would be formally more accurate indeed, since the Smolyak quadrature does not require to use the same quadrature rule for all dimensions. Notice that this complicates the mathematical equations. Hence, I think that the choice from (Gerstner & Griebel, 1992) is a reasonable trade-off and I suggest to stick to the current simplification.

**Comment by mbaudin47** [comment]
@adutfoy : I took into account for your comments in cb48d1134c717b3b6df2a6599cd213daba0a60e2. Are you ok with this PR?

File: python/src/GaussLegendre_doc.i.in [hunk 6]
```diff
@@ -9,34 +9,136 @@ Available constructors:
...
 ----------
 dimension : int, :math:`dimension>0`
-    The dimension of the functions to integrate. The default discretization is *GaussLegendre-DefaultMarginalIntegrationPointsNumber* in each dimension, see :class:`~openturns.ResourceMap`.
+    The input dimension of the functions to integrate.
 discretization : sequence of int
-    The number of nodes in each dimension. The sequence must be non-empty and must contain only positive values.
+    The number of nodes in each dimension.
+    The sequence must be non-empty and must contain only positive values.
+    The default discretization is *GaussLegendre-DefaultMarginalIntegrationPointsNumber*
+    in each dimension, see :class:`~openturns.ResourceMap`.
```
Comment (josephmure) [inline python/src/GaussLegendre_doc.i.in:17]:
```suggestion
    The default discretization is stored as the `GaussLegendre-DefaultMarginalIntegrationPointsNumber` :class:`~openturns.ResourceMap` key, which is used for each dimension.
```

File: python/src/GaussLegendre_doc.i.in [hunk 7]
```diff
@@ -9,34 +9,136 @@ Available constructors:
...
+The Gauss-Legendre algorithm approximates the definite integral:
 
 .. math::
 
-    \int_{\vect{a}}^\vect{b} f(\vect{t})\di{\vect{t}}
+    \int_{\vect{a}}^\vect{b} \model(\vect{x})\di{\vect{x}}
 
 
-with :math:`f: \Rset^d \mapsto \Rset^p`, :math:`\vect{a}, \vect{b}\in\Rset^d` using a fixed tensorized Gauss-Legendre approximation:
+where :math:`\model: \Rset^\inputDim \mapsto \Rset^\outputDim` is a function and
```
Comment (josephmure) [inline python/src/GaussLegendre_doc.i.in:28]: The changes to this file are in direct conflict with the changes proposed in PR https://github.com/openturns/openturns/pull/2766, you probably need to coordinate with @adutfoy

**Comment by adutfoy** [comment]
ok for me!

**Pull request merged**